framework/src/org/checkerframework/common/value/Range.java

```diff
@@ -0,0 +1,418 @@
+     * multiplication of two ranges.
+     *
+     * @param right the specified range to be multiplied by this range
+     * @return the range resulting from multiplying the specified range by this range
+     */
+    public Range times(Range right) {
+        List<Long> possibleValues =
+                Arrays.asList(from * right.from, from * right.to, to * right.from, to * right.to);
+        return new Range(Collections.min(possibleValues), Collections.max(possibleValues));
+    }
+
+    /**
+     * Returns the smallest range that includes all possible values resulting from dividing (integer
+     * division) an arbitrary value in this range by an arbitrary value in the specified range. We
+     * call this the division of two ranges.
+     *
+     * @param right the specified range by which this range is divided
+     * @return the range resulting from dividing this range by the specified range
+     */
+    public Range divide(Range right) {
+        long resultFrom = Long.MIN_VALUE;
+        long resultTo = Long.MAX_VALUE;
+
+        // Here we assume devide-by-zero is checked and avoided.
+        if (from > 0 && right.from >= 0) {
+            resultFrom = from / Math.max(right.to, 1);
+            resultTo = to / Math.max(right.from, 1);
+        } else if (from > 0 && right.to <= 0) {
+            resultFrom = to / Math.min(right.to, -1);
+            resultTo = from / Math.min(right.from, -1);
+        } else if (from > 0) {
+            resultFrom = -to;
+            resultTo = to;
+        } else if (to < 0 && right.from >= 0) {
+            resultFrom = from / Math.max(right.from, 1);
+            resultTo = to / Math.max(right.to, 1);
+        } else if (to < 0 && right.to <= 0) {
+            resultFrom = to / Math.min(right.from, -1);
+            resultTo = from / Math.min(right.to, -1);
+        } else if (to < 0) {
+            resultFrom = from;
+            resultTo = -from;
+        } else if (right.from >= 0) {
+            resultFrom = from / Math.max(right.from, 1);
+            resultTo = to / Math.max(right.from, 1);
+        } else if (right.to <= 0) {
+            resultFrom = to / Math.min(right.to, -1);
+            resultTo = from / Math.min(right.to, -1);
+        } else {
+            resultFrom = Math.min(from, -to);
+            resultTo = Math.max(-from, to);
+        }
+        return new Range(resultFrom, resultTo);
+    }
+
+    /**
+     * Returns a range that includes all possible values of the remainder of dividing an arbitrary
+     * value in this range by an arbitrary value in the specified range.
+     *
+     * @param right the specified range by which this range is divided
+     * @return the range of the remainder of dividing this range by the specified range. Note that
+     *     this range might not be the smallest range that includes all the possible values.
+     */
+    public Range remainder(Range right) {
+        ArrayList<Long> possibleValues =
```

Comment:
Unless there is some specific functionality of `ArrayList` that your code depends on, it's better to use a more general type when declaring a variable -- in this case, `List<Long>`.  That helps readers understand what the rest of the code depends on, and it insulates the code from implementation changes.